### PR TITLE
doc: add style guide

### PR DESCRIPTION
**Pull request by acud**
## Motivation

In an attempt to standardize how we check-in code to the codebase, I suggest we try to adopt a style-guide.

## Description

In order to bring in more uniformity and alignment in the codebase I am suggesting we make use of the Uber styleguide and modify it according to our needs. As in the `CODING.md` PR, this is meant as a departure point into a fruitful discussion. Please add your recommendations and comments.

## TODO

- [ ] add the stuff that needs addressing from the CODING.md PR

### CODE REVIEW

File: STYLE.md [hunk 1]
```diff
@@ -0,0 +1,4068 @@
...
+trips[0] = ...
+```
+
+</td>
+<td>
+
+```go
+func (d *Driver) SetTrips(trips []Trip) {
+  d.trips = make([]Trip, len(trips))
+  copy(d.trips, trips)
```
Comment (ivan4th) [inline STYLE.md]: There's [slices.Clone](https://pkg.go.dev/slices#Clone) for that now

File: STYLE.md [hunk 2]
```diff
@@ -0,0 +1,4068 @@
...
+3. [Go Code Review Comments](https://go.dev/wiki/CodeReviewComments)
+
+We aim for the code samples to be accurate for the two most recent minor versions
+of Go [releases](https://go.dev/doc/devel/release).
+
+All code should be error-free when run through `golint` and `go vet`. We
+recommend setting up your editor to:
+
+- Run `goimports` on save
+- Run `golint` and `go vet` to check for errors
```
Comment (fasmat) [inline STYLE.md]: We use more modern tools 🙂 

```suggestion
- Run `gofumpt` on save
- Run `make lint` or `golangci-lint` to check for errors
```

File: STYLE.md [hunk 3]
```diff
@@ -0,0 +1,4068 @@
...
+
+### Verify Interface Compliance
+
+Verify interface compliance at compile time where appropriate. This includes:
+
+- Exported types that are required to implement specific interfaces as part of
+  their API contract
+- Exported or unexported types that are part of a collection of types
+  implementing the same interface
+- Other cases where violating an interface would break users
```
Comment (fasmat) [inline STYLE.md:127]: I feel this should be more nuanced - yes if we implement an interface form the standard lib like `io.Writer` and `http.Handler` this makes sense. Those are interfaces that we have no control over and if we want to adhere to them a type check can be helpful to do so. However we often define our own interfaces, usually not even exporting them since they only exist for the purpose of decoupling / mocking. Generally these should be defined on the user side for a few reasons:

- prevents tight coupling between packages holding provider and user of implementation
- if interfaces / implementation is accidentally changed we still notice it right away on the user side
- if interfaces / implementation is intentionally changed its the same amount of effort either way - we generally only have one provider and few users - mocks can be regenerated

File: STYLE.md [hunk 4]
```diff
@@ -0,0 +1,4068 @@
...
+
+func (s *Stats) Snapshot() map[string]int {
+  s.mu.Lock()
+  defer s.mu.Unlock()
+
+  result := make(map[string]int, len(s.counters))
+  for k, v := range s.counters {
+    result[k] = v
+  }
+  return result
```
Comment (fasmat) [inline STYLE.md]:
```suggestion
  return slices.Clone(s.counters))
```

File: STYLE.md [hunk 5]
```diff
@@ -0,0 +1,4068 @@
...
+  but be aware that callers may begin to rely on this behavior.
+  So for cases where the wrapped error is a known `var` or type,
+  document and test it as part of your function's contract.
+- Use `%v` to obfuscate the underlying error.
+  Callers will be unable to match it,
+  but you can switch to `%w` in the future if needed.
+
+When adding context to returned errors, keep the context succinct by avoiding
+phrases like "failed to", which state the obvious and pile up as the error
+percolates up through the stack:
```
Comment (fasmat) [inline STYLE.md]: NIT: i.e. errors are not log messages! 🙂 👍

File: STYLE.md [hunk 6]
```diff
@@ -0,0 +1,4068 @@
...
+if !ok {
+  // handle the error gracefully
+}
+```
+
+</td></tr>
+</tbody></table>
+
+<!-- TODO: There are a few situations where the single assignment form is
+fine. -->
```
Comment (fasmat) [inline STYLE.md:1148]: Generally if you know the type assertion cannot fail:

- the type is the same: e.g. `type LayerID uint32` cast to `uint32` will aways work
- an `any` value is always of a certain type (and `any` is only used to satisfy some interface)
- etc.

File: STYLE.md [hunk 7]
```diff
@@ -0,0 +1,4068 @@
...
+something irrecoverable happens such as a nil dereference. An exception to this is
+program initialization: bad things at program startup that should abort the
+program may cause panic.
+
+```go
+var _statusTemplate = template.Must(template.New("name").Parse("_statusHTML"))
+```
+
+Even in tests, prefer `t.Fatal` or `t.FailNow` over panics to ensure that the
+test is marked as failed.
```
Comment (fasmat) [inline STYLE.md:1206]: Or the equivalent `assert.Fail`/`assert.FailNow` and `require.Fail`/`require.FailNow`

File: STYLE.md [hunk 8]
```diff
@@ -0,0 +1,4068 @@
...
+
+</td><td>
+
+```go
+// func TestFoo(t *testing.T)
+
+f, err := os.CreateTemp("", "test")
+if err != nil {
+  t.Fatal("failed to set up test")
+}
```
Comment (fasmat) [inline STYLE.md:1230]: Since we are using testify, I'd prefer:

```suggestion
require.NoError(t, err, "failed to set up test")
```

File: STYLE.md [hunk 9]
```diff
@@ -0,0 +1,4068 @@
...
+    wg.Add(1)
+    go func() {
+      defer wg.Done()
+      // ...
+    }()
+  }
+
+  // To wait for all to finish:
+  wg.Wait()
+  ```
```
Comment (fasmat) [inline STYLE.md:2046]: Or better: an `errgroup.Group`

File: STYLE.md [hunk 10]
```diff
@@ -0,0 +1,4068 @@
...
+</tbody></table>
+
+## Style
+
+### Avoid overly long lines
+
+Avoid lines of code that require readers to scroll horizontally
+or turn their heads too much.
+
+We recommend a soft line length limit of **99 characters**.
```
Comment (fasmat) [inline STYLE.md:2352]:
```suggestion
We recommend a soft line length limit of **120 characters**.
```

That's what our current linter enforces (must not be longer than that) and we should maybe reconsider increasing that limit. We are not in the 80 and still use terminals 😉. Even if I have 2 files open at the same time next to each other on a modern wide screen display there is still unused whitespace...

File: STYLE.md [hunk 11]
```diff
@@ -0,0 +1,4068 @@
...
+```
+
+</td><td>
+
+```go
+import (
+  "a"
+  "b"
+)
+```
```
Comment (fasmat) [inline STYLE.md]: For imports this is enforced by `golangci-lint` and can be autoformated with `gci` (or `golangci-lint --fix`) The order is:

```go
import (
    "io"
    "math" // stdlibrary

    "github.com/some/library"
    "github.com/stretchr/testify" // packages from other modules

    "github.com/spacemeshos/go-spacemesh/activation" // packages from same module
    "github.com/spacemeshos/go-spacemesh/proposal"
)
```

This is also enforced by `make lint`.

File: STYLE.md [hunk 12]
```diff
@@ -0,0 +1,4068 @@
...
+var (
+  a = 1
+  b = 2
+)
+
+type (
+  Area float64
+  Volume float64
+)
+```
```
Comment (fasmat) [inline STYLE.md]: This is actually enforced by `make lint`

File: STYLE.md [hunk 13]
```diff
@@ -0,0 +1,4068 @@
...
+</tbody></table>
+
+### Import Group Ordering
+
+There should be two import groups:
+
+- Standard library
+- Everything else
+
+This is the grouping applied by goimports by default.
```
Comment (fasmat) [inline STYLE.md]: See comment above

File: STYLE.md [hunk 14]
```diff
@@ -0,0 +1,4068 @@
...
+### Package Names
+
+When naming packages, choose a name that is:
+
+- All lower-case. No capitals or underscores.
+- Does not need to be renamed using named imports at most call sites.
+- Short and succinct. Remember that the name is identified in full at every call
+  site.
+- Not plural. For example, `net/url`, not `net/urls`.
+- Not "common", "util", "shared", or "lib". These are bad, uninformative names.
```
Comment (fasmat) [inline STYLE.md:2498]: ❤️

File: STYLE.md [hunk 15]
```diff
@@ -0,0 +1,4068 @@
...
+
+printInfo("foo", true /* isLocal */, true /* done */)
+```
+
+</td></tr>
+</tbody></table>
+
+Better yet, replace naked `bool` types with custom types for more readable and
+type-safe code. This allows more than just two states (true/false) for that
+parameter in the future.
```
Comment (fasmat) [inline STYLE.md:3192]: Even better: except for methods that only take a single bool as parameter, booleans generally indicate that a method/function should be two instead.

File: STYLE.md [hunk 16]
```diff
@@ -0,0 +1,4068 @@
...
+for _, tt := range tests {
+  // ...
+}
+```
+
+#### Avoid Unnecessary Complexity in Table Tests
+
+Table tests can be difficult to read and maintain if the subtests contain conditional
+assertions or other branching logic. Table tests should **NOT** be used whenever
+there needs to be complex or conditional logic inside subtests (i.e. complex logic inside the `for` loop).
```
Comment (fasmat) [inline STYLE.md]:
```suggestion
there needs to be complex or conditional logic inside subtests (i.e. complex logic inside a `for` loop
or changed behaviour of subtests due to an `if`).
```

File: STYLE.md [hunk 17]
```diff
@@ -0,0 +1,4068 @@
...
+}
+
+for _, tt := range tests {
+  tt := tt // for t.Parallel
+  t.Run(tt.give, func(t *testing.T) {
+    t.Parallel()
+    // ...
+  })
+}
+```
```
Comment (fasmat) [inline STYLE.md]: This has become obsolete with `Go 1.22` generally loop variables don't need to be copied any more even when using parallel execution.

File: STYLE.md [hunk 18]
```diff
@@ -0,0 +1,4068 @@
...
+### Lint Runners
+
+We recommend [golangci-lint](https://github.com/golangci/golangci-lint) as the go-to lint runner for Go code, largely due
+to its performance in larger codebases and ability to configure and use many
+canonical linters at once. This repo has an example [.golangci.yml](https://github.com/uber-go/guide/blob/master/.golangci.yml) config file
+with recommended linters and settings.
+
+golangci-lint has [various linters](https://golangci-lint.run/usage/linters/) available for use. The above linters are
+recommended as a base set, and we encourage teams to add any additional linters
+that make sense for their projects.
```
Comment (fasmat) [inline STYLE.md:3959]: We have a `.golangci-lint` configuration that we use for this project. Developers should use the configuration we use when linting their code for a contribution. 🙂

File: STYLE.md [hunk 19]
```diff
@@ -0,0 +1,4068 @@
...
+  if len(args) != 1 {
+    return errors.New("missing file")
+  }
+  name := args[0]
+
+  f, err := os.Open(name)
+  if err != nil {
+    return err
+  }
+  defer f.Close()
```
Comment (ivan4th) [inline STYLE.md:1843]: Note that this swallows error from `f.Close()` which is not good.
There's a pattern that helps with error handling in `defer`:  https://wstrm.dev/posts/errors-join-heart-defer/
Discussion on `/r/golang`: https://www.reddit.com/r/golang/comments/11luvxm/errorsjoin_defer/
Also: https://stackoverflow.com/a/76287159
IMO it has a downside, namely that you must not forget to use named return value (which by itself is sometimes frowned upon when you don't specify the actual return values in `return`), but still it's probably better than alternatives

```go
func run() (err error) {
  args := os.Args[1:]
  if len(args) != 1 {
    return errors.New("missing file")
  }
  name := args[0]

  f, err := os.Open(name)
  if err != nil {
    return err
  }
  defer func () {
    err = errors.Join(err, f.Close())
  }()

  b, err := io.ReadAll(f)
  if err != nil {
    return err
  }

  // ...
}
```

If we're to use this pattern, it probably belongs to the section on `defer`

Comment (fasmat) [reply STYLE.md:1843]: I'm not sure if I like this pattern. It doesn't handle the error. If closing fails this should in my opinion be handled where the closing happened and not further up the call stack. This pattern only achieves that the error isn't lost.

Also this introduces possible false positives:

```go
func run() (err error) {
	f, _ := os.Create("test.txt")
	defer func() {
		err = errors.Join(err, f.Close())
	}()

	f.Write([]byte("test"))
	if err := f.Close(); err != nil { // closed explicitly to handle error
		// handle err instead of just logging it
		log.Fatal(err)
	}
	return nil
}

func main() {
	fmt.Println(run())
}
```

This prints the error "file closed already" which isn't an error and should be filtered, making the `defer`red function unnecessarily complex 😕

Comment (ivan4th) [reply STYLE.md:1843]: The problem is that if we have many clauses like this in `run()`
```go
  b, err := io.ReadAll(f)
  if err != nil {
    return err
  }

  if err := doSmthWithBuffer(b); err != nil {
    return fmt.Errorf("smth: %w", err)
  }

  if err := doSmthElse(b); err != nil {
    return fmt.Errorf("smth else: %w", err)
  }
```
we would have to rewrite them like
```go
  b, err := io.ReadAll(f)
  if err != nil {
    return errors.Join(err, f.Close())
  }

  if err := doSmthWithBuffer(b); err != nil {
    return errors.Join(fmt.Errorf("smth: %w", err), f.Close())
  }

  if err := doSmthElse(b); err != nil {
    return errors.Join(fmt.Errorf("smth else: %w", err), f.Close())
  }
```
which quickly becomes error-prone, as it's easy to shoot yourself in the foot by missing `f.Close()` in one of the branches as the code evolves. Moreover, if instead of just closing the file we had some more advanced cleanup that must run even if the code panics somewhere, it will fail to run.
Basically that's the situation for which `defer` exists, but we also need to handle errors that happen during the cleanup. Good if we can just log them and that's it, but such an approach is not always applicable.

Of course care must be taken not to close the file twice and not to handle the error twice (log and return), so in some cases the `defer` approach is not appropriate. But if you need to always close the file at the end of the function, this works quite well as you don't miss any errors and still have benefits of `defer`.

EDIT: probably `io.ReadAll()` is not the best example to illustrate my point, as in most cases you can close the file immediately following the call, but let's assume `f` still needs to be open after that point

Comment (fasmat) [reply STYLE.md:1843]: But you are not handling the error, you are just appending it to another - it might even be caused by the first error to begin with?

Instead of an anonymous deferred function the closing should be handled properly - if it even makes a difference to begin with 🤷 

```go
func closeWithErrorHandling(f io.Closer) {
    err := f.close()
    if err == nil {
        return
    }
    // handle error
}

func run() error {
    f, _ := os.Create("test.txt")
    defer closeWithErrorHandling(f)

    // proceed doing with `f` what needs to be done
}

Comment (ivan4th) [reply STYLE.md:1843]: This kind of error handling will not cause `run()` to return an error, and oftentimes we might want to indicate that `run` has failed to the caller instead of (or in addition to) doing any special handling in the `run`.
Of course if `run` can do all the handling we need and then return `nil` error such as log the error this pattern is unnecessary, but that is not what we always need.

So, just to be clear, the circumstances that warrant the use of the pattern are:
* a function needs to perform some cleanup when returning
* the cleanup may fail
* there are multiple `return` points in the function upon which the cleanup must be performed
* the function cannot perform all the necessary error handling on its own and must return an error if cleanup fails

This may sound a bit too specific but it's actually rather common situation

Comment (fasmat) [reply STYLE.md:1843]: but `run` will return an error if it's a different one from the one caused by `Close`? The error from close doesn't need to be returned it has already been handled...

EDIT: also if another error occurs the error from the deferred function isn't helpful in the investigation of what went wrong usually. So the error only needs to be handled if everything else went well.

I think this need to handle an error in a deferred function rarely if ever comes up if the cleanup is called explicitly on the good path and will usually just lead to polluting the actual issue with noise.

Comment (ivan4th) [reply STYLE.md:1843]: Whether returning errors in `defer` in case of failure is counterproductive is debatable, depending on the situation at hand.
E.g. if we're closing the database after a failed query, normally closing the database will succeed if the problem is a simple SQL error, but if the problem has happened due to some severe database corruption, failed `Close` may give some additional hints. IoW: in the majority of cases, the cleanup should succeed even on the "unhappy" path, and in rare cases where it fails the cleanup failure may provide additional hints.

Besides, a function may have several happy paths with early returns in addition to error returns, and upon such "happy" returns we may want to perform the cleanup and let the caller know if smth went wrong at the cleanup phase, too.

File: STYLE.md (hunk 3 above)
Comment (acud) [reply STYLE.md:127]: I feel that having the unexported interfaces is a bit skewed in the first place. 

Ideally those interfaces should be exposed already publicly on the dependency package.

For example, if the p2p server exposes some pubsub methods through an interface, it should also generate the p2p mocks and we should consume those everywhere in the codebase where this dependency is used (instead of duplicating those interfaces endlessly, and generating redundant mocks). It generates less points of friction for testing and for handling the dependency injection.

In some cases indeed subsets of interfaces are needed but I'm not sure why not to just compose the interfaces together at the dependency level, this is not unheard of.

Another point for public interfaces is allowing us to properly do black-box testing which I feel is way superior to white-box. Ideally, white-box should be resorted to only in exceptional cases, but otherwise tests shouldn't leak package internals. Currently most of the tests don't have namespace separation and I feel it causes a lot of mess when asserting state. A lot of the necessary helpers can be put in an `export_test.go` [file](https://medium.com/@robiplus/golang-trick-export-for-test-aa16cbd7b8cd) (which is a bit hacky in the first place, but at least puts a lid on hackiness).

Comment (acud) [reply STYLE.md:127]: And back to the conversation about interface compliance - it would be nice to have, but we could also change it to compliance with external interfaces only for now, it is fine by me.

File: STYLE.md (hunk 4 above)
Comment (acud) [reply STYLE.md]: you mean `maps.Clone`?

File: STYLE.md (hunk 7 above)
Comment (acud) [reply STYLE.md:1206]: I would personally stick to the `t.Fatal`/`t.FailNow`. I'm not sure what value do these external packages provide to such a basic operation.

File: STYLE.md (hunk 8 above)
Comment (acud) [reply STYLE.md:1230]: Can you please elaborate on why this is needed? I am at the opinion that this deviates from the general style-guide and personally makes things harder to read. Having this sort of deviation from the idiomatic go error handling blocks encourages all sorts of schtiks that reduce code readability in favor of developer creativity and indirection, etc. I don't think this is necessary and that we should start being lazy about writing explicit error handling blocks.

File: STYLE.md (hunk 7 above)
Comment (fasmat) [reply STYLE.md:1206]: We are already using them in tests to a high degree, the testify versions call `t.Fatal`/`t.FailNow` but add more context to the error message and print them in a visually appealing way 🙂

File: STYLE.md (hunk 10 above)
Comment (acud) [reply STYLE.md:2352]: Yeah I totally agree. I am also on `neovim` with a 13 something inch screen and I find the 120 character to be extremely limiting and makes a lot of method calls to be very strange to read with all the line breaks. Not sure why this is needed to be so small of a value.

File: STYLE.md (hunk 6 above)
Comment (acud) [reply STYLE.md:1148]: Added some comment about this. :+1:

File: STYLE.md (hunk 8 above)
Comment (fasmat) [reply STYLE.md:1230]: Lets use a different example:

```go
require.ErrorIs(t, err, io.EOF)
```

this is much more succinct than:

```go
if err == nil {
    t.Fatal("expected error")
}
if !errors.Is(err, io.EOF) {
   t.Fatal("didn't receive expected error; got %T", err)
}
```

besides this doesn't print all the information `require.ErrorIs` would print.

File: STYLE.md (hunk 11 above)
Comment (acud) [reply STYLE.md]: So should we remove the entry for this? (in case it is already enforced by static analysis)

File: STYLE.md (hunk 12 above)
Comment (acud) [reply STYLE.md]: Should remove?

File: STYLE.md (hunk 13 above)
Comment (acud) [reply STYLE.md]: ditto

File: STYLE.md (hunk 15 above)
Comment (acud) [reply STYLE.md:3192]: Can you give an example just for the sake of clarity?

File: STYLE.md (hunk 18 above)
Comment (acud) [reply STYLE.md:3959]: Added as a comment :+1:

File: STYLE.md (hunk 15 above)
Comment (fasmat) [reply STYLE.md:3192]:
```go
func CalcRank(wins int, losses int, isTournament bool) {
    if(isTournament) {
        return wins / (wins + losses)
    }
    
    return (wins) / (losses)
}
```

should be 2 functions: `CalcRank` and `CalcTournamentRank`. There are (rare) exceptions, but usually a boolean indicates that it should be two methods/functions instead.

File: STYLE.md (hunk 3 above)
Comment (fasmat) [reply STYLE.md:127]: The implementation check `var _ MyInterface = MyType{}` doesn't add anything to our codebase, also without it we quickly notice that a type that should implement an interface doesn't any more because at the point of its use it will cause a compilation error.

We are not providing a library to be used by others; the interfaces we use exist for decoupling and mocking. The reason we define interfaces at the user side is because the user decides which subset of the service they want to use and/or mock.

If the implementation changes its signature we have to update all uses anyway, doesn't matter if the interface is on the user or the provider side. I don't mind trying more aggressively to go to black box tests, but even then I see no need to define interfaces at the implementation side - we want to avoid dependencies between packages. Ideally only `node` has a dependency on (almost) everything while all other packages have dependencies on (almost) nothing.

We are right now busy with refactorings because interfaces were defined on the implementation side: In the `malfeasance` package used to be (and still partially is) the interface and types for all packages creating malfeasance proofs, but those proofs contain types of those other packages:

- `activation` creates a `malfeasance.MalfeasanceProof`, inside there is a `activation.ActivationTx`
- we now have a cyclic dependency between `activation` and `malfeasance`
- the "solution" was to create more packages: `malfeasance` depends on `malfeasance/wire`, `activation/wire` and `activation` depends on `malfeasance`, `malfeasance/wire` and `activation/wire`
- its now 4 packages instead of 2 just to get around the issue of that activation depends on malfeasance for publishing proofs and malfeasance depends on activation to validate them

File: STYLE.md (hunk 19 above)
Comment (acud) [reply STYLE.md:1843]: I'd like to chip in here because this showed up in your PR @ivan4th and I also commented on this.

I am at the opinion that if we have to add this sort of cleanup code that has all sorts of complicated error instrumentation that is just difficult to understand and is not so informative (how do you know which line in the error corresponds to which `errors.Join`, etc), it is probably cleaner and nicer to do the cleanup _once_ on the calling context (i.e. the context that called the function that has to do the convoluted cleanup logic). Having these sort of anonymous functions is really difficult to read (to whoever didn't write them usually).

As for `errors.Join`, I find it useful in a case of a task that is parallelized (like in an `errgroup`, but then again `errgroup` should do this for you anyway a-la-carte).

On a personal note: I've seen maybe a case or two  where this sort of cleanup logic was needed so far (not in this project) and it was in a p2p protocol, where developer error was very easy to do and have caused certain cases where a `return` statement could miss satisfying all sorts of other calls that had to be done (cleanup of the libp2p stream, logging, metrics, etc), which is why it was later decided to have an anonymous function that catches whether an error was caught, causing the cleanup to be made more nicely (however, the code-smell is that the protocol was already way to convoluted, and this is why this was needed in the first place).

My suggestion here is that maybe we could also do a pair/mob programming session on the code in question (in the sql vaacum PR) so that we could see together whether the `errors.Join` is needed or not.

File: STYLE.md (hunk 3 above)
Comment (acud) [reply STYLE.md:127]: Ok point made :) I will amend

File: STYLE.md (hunk 19 above)
Comment (ivan4th) [reply STYLE.md:1843]: I'm not insisting on the use of said pattern, if others don't like it, I will not use it, and I will update vacuum PR accordingly :)
If smth like that is to be used, simple `errors.Join` is probably indeed not the best approach, a cleanup errors need to be augmented, too, to make it cleaner what really happened. 

Moving cleanup to the calling context may often make sense, but in some cases it may also cause proliferation of funcs like `doSomethingInner`, `doInnerSomethingOfAnotherInnerSomething`, etc, if the only purpose of extracting a part of function is making cleanup easier. 

But in any case, simple `defer f.Close()` in the original example can be said to constitute a bug, as it may swallow a real I/O error which happens when the file is closed.

Comment (ivan4th) [reply STYLE.md:1843]: Re pair/mob programming session, that might be a good idea, too

File: STYLE.md [hunk 20]
```diff
@@ -0,0 +1,4068 @@
...
+}
+```
+
+Note that there's a method of implementing this pattern with closures but we
+believe that the pattern above provides more flexibility for authors and is
+easier to debug and test for users. In particular, it allows options to be
+compared against each other in tests and mocks, versus closures where this is
+impossible. Further, it lets options implement other interfaces, including
+`fmt.Stringer` which allows for user-readable string representations of the
+options.
```
Comment (ivan4th) [inline STYLE.md:3925]: As it was mentioned by @acud elsewhere, overusing options as we do it now may not always be good, so maybe some discussion needed regarding the alternatives, and some examples showing when to use options and when to use a simple approach.
Another problem is that the approach to options which is described here differs from what's in use in the most of go-spacemesh codebase at this moment, which is using closure-based options. Although the advantages described in this paragraph may be valid, they may not be always worth the cost of adding verbosity to already verbose options mechanism (which is essentially a verbose kwargs replacement), and also in this case we will either have inconsistent option mechanisms in the codebase, or we will have to refactor a lot of existing code.

File: STYLE.md (hunk 19 above)
Comment (fasmat) [reply STYLE.md:1843]: > But in any case, simple defer f.Close() in the original example can be said to constitute a bug, as it may swallow a real I/O error which happens when the file is closed.

Yes this is true but if that happens, it doesn't really matter that `f.Cose()` fails:

```go
f, err := os.Open("path/to/file")
if err != nil {
    return err
}
defer f.Close()
buf, err := io.ReadAll(f)
if err != nil {
    return fmt.Error("read file: %w", err) // if close fails it doesn't really matter; read already failed, close will fail for the same cause if it does
}
// do something with `buf`
// happy case reached we now close explicitly
if err := f.Close(); err != nil {
    return fmt.Errorf("closing file: %w", err)
}
return nil
```

File: STYLE.md (hunk 20 above)
Comment (fasmat) [reply STYLE.md:3925]: The difference between "closure options" and "functional options" is only that the closure options manipulate the created object directly, while the functional operate on an intermediary object that only holds configuration and than use that to construct the actual service. Sure in the details it is different but from the point of using them I see little difference.

Functional options should be used for _optional_ parameters and thats where they are currently overused in our codebase. In many places the parameters that are passed to the services are not optional and not setting them will cause the instantiation of the object to fail.

I've occasionally done this as well primarily for readability:

```go
func NewService(
    db,
    lg,
    clk,
    sync,
    h,
)
```

is just much less readable than

```go
func NewService(
    WithDB(db),
    WithLogger(lg),
    WithClock(clk),
    WithSyncer(sync),
    WithHandler(h),
)
```

Sure variable naming can be improved too to mitigate this issue but functional options have an additional benefit for testing where constructing many services with slightly different configuration becomes much easier:

```go
defaultOpts := []Opts{
    WithDB(db),
    WithLogger(lg),
    WithClock(clk),
    WithSyncer(sync),
    WithHandler(h),
}

testservice1 := NewService(defaultOpts..., WithHandler(mockHandler))
testservice2 := NewService(defaultOpts..., WithSyncer(mockSyncer))
// ... etc.
```

File: STYLE.md (hunk 19 above)
Comment (ivan4th) [reply STYLE.md:1843]: @fasmat my point is, the original code doesn't have any handling for `f.Close()` errors at all:
```go

  f, err := os.Open(name)
  if err != nil {
    log.Fatal(err)
  }
  defer f.Close()

  // If we call log.Fatal after this line,
  // f.Close will not be called.

  b, err := io.ReadAll(f)
  if err != nil {
    log.Fatal(err)
  }
```
Granted, we're reading the file not writing it, which kind of makes implications less severe, but still completely ignore `Close` errors is not correct. In this case, the `Close` error will be ignored even if `io.ReadAll` succeeds.

File: STYLE.md (hunk 20 above)
Comment (ivan4th) [reply STYLE.md:3925]: Problem is that the paragraph I commented recommends against using closure-based options. Thus we either update the guide, refactor our code or accept it as an inconsistency

As of options being useful in many cases, I can only agree with that, but there are also arguments against using them too much:
https://github.com/spacemeshos/go-spacemesh/pull/6085#discussion_r1660225344

I for one don't find these arguments to be applicable to all cases as having 10 different constructors to avoid the use of the option pattern doesn't sound too good to me, but there probably needs to be some kind of balance

Comment (acud) [reply STYLE.md:3925]: So I'm with @ivan4th on this. Personally I find this sort of pattern to be very confusing because it is also allowing for irregular sequence of setting things in a constructor. It has the sort of flexibility that the language intentionally tries to avoid for various reasons.
Also having these sort of parameters has the unfortunate consequence that a lot of the components in the code-base do things in a very complex way. For example:
- component defines it's _own_ configuration with a default configuration which is then _potentially_ overridden by the instantiator (BAD) - configuration should never be an optional parameter
- logger is defined by the component but then _potentially_ overridden by the instantiator - why should a logger parameter be optional? for me this is a mandatory thing and that instantiator should know whether the logger that should be passed is a noop logger or not
- DB is not optional and should not be marked as such by using this pattern.

IMO the "internal vocalisation" of how we read code shouldn't affect the way we write it. And yes I agree that sometimes having a constructor call with a lot of arguments can become cumbersome but you at least know for sure what is needed and it usually also breaks-fast so one can fix it fast.

So TL;DR my personal opinion is that I find this pattern less readable and that it adds an unreasonable amount of mental taxation reading the code (and resolving its dependencies)

File: STYLE.md (hunk 19 above)
Comment (acud) [reply STYLE.md:1843]: So I think that @fasmat is arguing here that handling the `Close` doesn't really matter because probably the error is so severe that the application will exit anyway. From my personal experience, having a `db.Close` on some part of a component that is going to raise and exception anyway is _best-effort_ at best. If you can't `Close` a db it usually means that:
- you have some very severe I/O errors on the hardware level
- disk is out of space and you can't flush stuff to it anymore

In our problem domain, I believe that other issues are probable but I'm more familiar with the ones above. Mostly, in those cases you assume that the node is no longer able to remain consistent (once you're out of space you're screwed because you don't know whether a "large" write is successful vs "small" write which largely leaves you with the assumption of no-longer-consistent).

In those cases node-operator is needed to recover the node and usually you need to start over. Hence I support that `Close` is a best-effort and the error can be ignored. 

https://stackoverflow.com/questions/37977714/can-i-get-an-error-when-closing-a-file

**Comment by acud** [comment]
Closing this since it doesn't seem to be relevant or important enough to complete
